system_test/helpers/aest_docker.erl

```diff
@@ -0,0 +1,260 @@
+        volumes => [
+            {ro, ConfigFilePath, ?EPOCH_CONFIG_FILE},
+            {rw, LogPath, ?EPOCH_LOG_FOLDER}
+        ],
+        ports => PortMapping
+    },
+    #{'Id' := ContId} = aest_docker_api:create_container(Hostname, DockerConfig),
+    log(NodeState, "Container ~p [~s] created", [Name, ContId]),
+    NodeState#{
+        container_name => Hostname,
+        container_id => ContId,
+        config_path => ConfigFilePath
+    }.
+
+
+-spec delete_node(NodeState) -> ok
+    when NodeState :: node_state().
+
+delete_node(#{container_id := ID, hostname := Name} = NodeState) ->
+    aest_docker_api:delete_container(ID),
+    log(NodeState, "Container ~p [~s] deleted", [Name, ID]),
+    ok.
+
+
+-spec start_node(NodeState) -> NodeState
+    when NodeState :: node_state().
+
+start_node(#{container_id := ID, hostname := Name} = NodeState) ->
+    aest_docker_api:start_container(ID),
+    log(NodeState, "Container ~p [~s] started", [Name, ID]),
+    NodeState.
+
+
+-spec stop_node(NodeState) -> NodeState
+    when NodeState :: node_state().
+
+stop_node(NodeState) -> stop_node(NodeState, #{}).
+
+
+-spec stop_node(NodeState, Options) -> NodeState
+    when NodeState :: node_state(),
+         Options :: #{
+            soft_timeout => pos_integer(),
+            hard_timeout => pos_integer()
+         }.
+
+stop_node(#{container_id := ID, hostname := Name} = NodeState, Opts) ->
+    aest_docker_api:stop_container(ID, Opts),
+    log(NodeState, "Container ~p [~s] stopped", [Name, ID]),
+    NodeState.
+
+
+-spec kill_node(NodeState) -> NodeState
+    when NodeState :: node_state().
+
+kill_node(#{container_id := ID, hostname := Name} = NodeState) ->
+    aest_docker_api:kill_container(ID),
+    log(NodeState, "Container ~p [~s] killed", [Name, ID]),
+    NodeState.
+
+
+-spec get_service_address(NodeState, Service) -> binary()
+    when NodeState :: node_state(), Service :: service_label().
+
+get_service_address(Service, NodeState) ->
+    #{local_ports := #{Service := Port}} = NodeState,
+    format("http://localhost:~w/", [Port]).
+
+%=== INTERNAL FUNCTIONS ========================================================
+
+log(#{log_fun := LogFun}, Fmt, Args) -> log(LogFun, Fmt, Args);
+log(undefined, _Fmt, _Args) -> ok;
+log(LogFun, Fmt, Args) when is_function(LogFun) -> LogFun(Fmt, Args).
+
+uid2postfix(undefined) -> <<>>;
+uid2postfix(<<>>) -> <<>>;
+uid2postfix(Uid) -> <<"_", Uid/binary>>.
+
+free_port() ->
+    {ok, Socket} = gen_tcp:listen(0, [{reuseaddr, true}]),
+    {ok, Port} = inet:port(Socket),
+    gen_tcp:close(Socket),
+    Port.
+
+allocate_ports(Labels) -> allocate_ports(Labels, #{}).
+
+allocate_ports([], Acc) -> Acc;
+allocate_ports([Label | Labels], Acc) ->
+    allocate_ports(Labels, Acc#{Label => free_port()}).
+
+
+format(Fmt, Args) ->
+    iolist_to_binary(io_lib:format(Fmt, Args)).
+
+write_template(TemplateFile, OutputFile, Context) ->
+    Template = bbmustache:parse_file(TemplateFile),
```

Comment:
It looks like this handles unexistent file silently. Not blocker for PR though.

I put together patch for making this more strict https://github.com/lucafavatella/epoch/commit/df66cc92637cea88227ef1d1c436d4b882694891
After this PR is merged to master I plan to open PR proposing this (though feel free to include in this if your want).

Reply:
I'll include it.